File: FacebookInterviewQuestions/verticalOrderTraversalOfABinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalTraversal = function(root) {
  let map = new Map();
  let queue = [];
  root.dist = 0;
  queue.push(root);
  let min = 0;
  let max = 0;
  while (queue.length > 0) {
    let n = queue.length;
    let temp = new Map();
    for (let i = 0; i < n; i++) {
      let node = queue.shift();
      if (!temp.has(node.dist)) {
        temp.set(node.dist, []);
      }
      let arr = temp.get(node.dist);
      arr.push(node.val);
      temp.set(node.dist, arr);
      if (node.left !== null) {
        node.left.dist = node.dist - 1;
        min = Math.min(node.left.dist, min);
        queue.push(node.left);
      }
      if (node.right !== null) {
        node.right.dist = node.dist + 1;
        max = Math.max(node.right.dist, max);
        queue.push(node.right);
      }
    }
    for (let key of temp.keys()) {
      if (!map.has(key)) {
        map.set(key, []);
      }

      let t = map.get(key);
      let val = temp.get(key);
      val.sort((a, b) => a - b);
      t.push(...val);

      map.set(key, t);
    }
  }

  let result = [];
  //   console.log(min, max);

  for (let i = min; i <= max; i++) {
    let temp = map.get(i);
    result.push(temp);
  }
  return result;
};
class Node {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

let node = new Node(3);
node.left = new Node(9);
node.right = new Node(20);
node.right.left = new Node(15);
node.right.right = new Node(7);
console.log(verticalTraversal(node));

File: BloombergInterviewQuestions/verticalOrderTraversalOfABinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class XTreeNode(TreeNode):
    def __init__(self, node):
        TreeNode.__init__(self, node.val, node.left, node.right)
        self.dist = 0
class Solution:
    def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        
        store = {}
        
        ans = []
        queue = []
        queue.append(XTreeNode(root))
        min_val = 0
        max_val = 0
        while queue:
            n = len(queue)
            tmp = {}
            for i in range(n):

                node = queue.pop(0)
                if node.dist in tmp:
                    tmp[node.dist].append(node.val)
                else:
                    tmp[node.dist] = [node.val]

                if node.left:
                    node.left = XTreeNode(node.left)
                    node.left.dist = node.dist - 1
                    min_val = min(min_val, node.dist - 1)
                    queue.append(node.left)

                if node.right:
                    node.right = XTreeNode(node.right)
                    node.right.dist = node.dist + 1
                    max_val = max(max_val, node.dist + 1)
                    queue.append(node.right)
                    
            for key in tmp:
                if key not in store:
                    store[key] = []
                arr = sorted(tmp[key])
                store[key].extend(arr)
    
        # print(store)
        for k in range(min_val, max_val + 1):
            tmp = store[k]
            ans.append(tmp)
            
        return ans